IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,466 @@
+    query[(id)kSecClass] = (id)kSecClassGenericPassword;
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    query[(id)kSecReturnAttributes] = @YES;
+
+    NSArray *items;
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, (void *)&items);
+    if (status == errSecItemNotFound) {
+        return @[];
+    } else if (status != errSecSuccess) {
+        NSString *errorMessage = @"Failed to read account from keychain";
+        MSID_LOG_WARN(context, @"%@ (%d)", errorMessage, status);
+        if (error) {
+            *error = MSIDCreateError(
+                NSOSStatusErrorDomain, (NSInteger)status, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return nil;
+    }
+
+    NSMutableArray<MSIDAccountCacheItem *> *accountList = [NSMutableArray new];
+    for (NSDictionary *itemDict in items) {
+        if ([self accountItem:itemDict matchesKey:key]) {
+            // This item matches the key, so try to deserialize it and allocate an account object for it
+            NSMutableDictionary *itemQuery = [self defaultAccountQuery:nil];
+            itemQuery[(id)kSecAttrAccount] = itemDict[(id)kSecAttrAccount];
+            itemQuery[(id)kSecReturnData] = @YES;
+            NSData *jsonData;
+            status = SecItemCopyMatching((CFDictionaryRef)itemQuery, (void *)&jsonData);
+            if (status != errSecSuccess) {
+                if (error) {
+                    NSString *errorMessage = @"Failed to read account from keychain";
+                    MSID_LOG_WARN(context, @"%@ (%d)", errorMessage, status);
+                    if (error) {
+                        *error = MSIDCreateError(
+                            NSOSStatusErrorDomain, (NSInteger)status, errorMessage, nil, nil, nil, context.correlationId, nil);
+                    }
+                }
+                continue;
+            }
+            if (jsonData) {
+                MSIDAccountCacheItem *account =
+                    (MSIDAccountCacheItem *)[serializer deserializeAccountCacheItem:jsonData];
+                if (account == nil) {
+                    NSString *errorMessage = @"Failed to deserialize account";
+                    *error = MSIDCreateError(
+                        MSIDErrorDomain, (NSInteger)MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil);
+                    MSID_LOG_WARN(context, @"%@", errorMessage);
+                    continue;
+                }
+                [accountList addObject:account];
+            }
+        }
+    }
+
+    return accountList;
+}
+
+// Remove one or more accounts from the keychain that match the key (see accountItem:matchesKey).
+- (BOOL)removeItemsWithAccountKey:(MSIDCacheKey *)key
+                          context:(__unused id<MSIDRequestContext>)context
+                            error:(NSError **)error {
+    MSID_TRACE;
+    NSMutableDictionary *query = [NSMutableDictionary new];
```

Comment:
I don't think we should be reading all items. Instead, we should be using similar model to what iOS keychain implementation is doing by providing maximum key available and filtering items on top level.

Reply:
This now follows the iOS approach. The higher level code on our side that was using this has been revised to no longer need it.